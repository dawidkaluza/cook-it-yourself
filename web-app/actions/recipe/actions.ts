"use server"

type Recipes = {
  items: [
    id: number,
    name: string,
    description: string
  ]
};

export const getMyRecipes = async () : Promise<Recipes> => {

};
