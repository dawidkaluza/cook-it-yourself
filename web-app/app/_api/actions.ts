"use server"

import {cookies} from "next/headers";

export function isSignedIn(){
  const cookieStore = cookies();
  const sessionCookie = process.env.NEXT_PUBLIC_API_GATEWAY_SESSION_COOKIE;
  return sessionCookie ? cookieStore.has(sessionCookie) : false;
}