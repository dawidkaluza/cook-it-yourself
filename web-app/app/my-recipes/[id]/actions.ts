"use server";

import {RecipeDetails} from "@/app/my-recipes/_dtos/recipe";
import { fetchFromServer} from "@/app/_api/fetch";

export async function reviewRecipe(id: number) : Promise<RecipeDetails> {
  return await fetchFromServer<RecipeDetails>({
    endpoint: `/kitchen/recipe/${id}`
  }) as RecipeDetails;
}