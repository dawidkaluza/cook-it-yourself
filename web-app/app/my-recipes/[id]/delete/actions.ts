"use server";

import {ApiError, fetchFromServer} from "@/app/_api/fetch";
import {redirect} from "next/navigation";

export async function deleteRecipe(prevState: any, formData: FormData): Promise<string> {
  const id = formData.get("id");

  try {
    await fetchFromServer({
      endpoint: `/kitchen/recipe/${id}`,
      method: "DELETE"
    });
  } catch (error) {
    if (error instanceof ApiError) {
      const response = error.response;
      switch (response.status) {
        case 404: {
          return "The recipe is already deleted.";
        }

        case 422: {
          return "Couldn't process the request.";
        }
      }
    }

    throw error;
  }

  redirect("/my-recipes");
}