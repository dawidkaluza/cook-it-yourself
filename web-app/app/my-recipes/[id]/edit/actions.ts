"use server";

import {FieldError} from "@/app/my-recipes/_dtos/errors";
import {ApiError, fetchFromServer} from "@/app/_api/fetch";
import {PortionSize, RecipeDetails, UpdateRecipeRequest} from "@/app/my-recipes/_dtos/recipe";
import {redirect} from "next/navigation";

async function editRecipe(prevState: any, formData: FormData): Promise<FieldError[]> {
  const id = formData.get("id") as string;

  const mapToPortionSize = (formData: FormData): PortionSize => {
    const portionSizeInput = formData?.get("portionSize")?.toString();
    if (!portionSizeInput) {
      return {
        value: "",
        measure: "",
      };
    }

    const numberIndex = portionSizeInput.search(/\d/g);
    if (numberIndex === -1) {
      return {
        value: "",
        measure: "",
      };
    }

    const nonNumberIndex = portionSizeInput.search(/\D/g);
    if (nonNumberIndex === -1 || nonNumberIndex < numberIndex) {
      return {
        value: portionSizeInput.slice(numberIndex),
        measure: "",
      };
    }

    return {
      value: portionSizeInput.slice(numberIndex, nonNumberIndex).trim(),
      measure: portionSizeInput.slice(nonNumberIndex).trim()
    };
  }

  const requestData: UpdateRecipeRequest = {
    basicInformation: {
      name: formData.get("name") as string,
      description: formData.get("description") as string,
      cookingTime: Number(formData.get("cookingTime")),
      portionSize: mapToPortionSize(formData),
    },
    ingredients: {
      ingredientsToAdd: [],
      ingredientsToUpdate: [],
      ingredientsToDelete: [],
    },
    steps: {
      stepsToAdd: [],
      stepsToUpdate: [],
      stepsToDelete: [],
    }
  };

  try {
    const updatedRecipe = await fetchFromServer<RecipeDetails>({
      endpoint: `/kitchen/recipe/${id}`,
      method: "PUT",
      body: requestData
    });

    if (updatedRecipe) {
      redirect(`/my-recipes/${updatedRecipe.id}/`);
    }
  } catch (error) {
    if (error instanceof ApiError) {
      const response = error.response;

      // TODO cover other possible client errors
      if (response.status === 422) {
        const body = await response.json();
        const fields : FieldError[] = body.fields;
        return fields.map((field) => {
          const name = field.name;
          const message = field.message;
          const indexOfDot = name.indexOf(".");
          if (indexOfDot === -1) {
            return {
              name,
              message
            }
          } else {
            const mainName = name.slice(0, indexOfDot);
            return {
              name: mainName,
              message
            };
          }
        });
      }

      return [];
    }

    throw error;
  }

  throw new Error("Couldn't process the request");
}

export { editRecipe };