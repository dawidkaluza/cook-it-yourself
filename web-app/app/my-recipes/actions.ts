"use server";

import {fetchFromServer} from "@/app/_api/fetch";
import {Page} from "@/app/my-recipes/_dtos/page";
import {Recipe} from "@/app/my-recipes/_dtos/recipe";

export async function getMyRecipes(){
  return await fetchFromServer({ endpoint: "/kitchen/recipe" }) as Promise<Page<Recipe>>;
}