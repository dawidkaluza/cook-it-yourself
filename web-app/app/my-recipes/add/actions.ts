"use server";

import {Ingredient, PortionSize, Recipe, Step} from "@/app/my-recipes/add/_dtos/Recipe";
import {ApiError, fetchFromServer} from "@/app/_api/fetch";
import {redirect} from "next/navigation";

export type FieldError = {
  name: string;
  message: string;
};

export async function addRecipe(prevState: any, formData: FormData) : Promise<FieldError[]> {
  const mapToIngredients = (formData: FormData) : Ingredient[] => {
    const ingredientsNames  = formData.getAll("ingredientName") as string[];
    const ingredientsAmounts = formData.getAll("ingredientAmount") as string[];
    const ingredientsUnits = formData.getAll("ingredientUnit") as string[];
    const ingredients : Ingredient[] = [];
    const ingredientsNum = ingredientsNames.length;
    for (let i = 0; i < ingredientsNum; i++) {
      const ingredient = {
        name: ingredientsNames[i],
        value: ingredientsAmounts[i],
        measure: ingredientsUnits[i],
      };
      ingredients.push(ingredient);
    }
    return ingredients;
  };

  const mapToSteps = (formData: FormData) : Step[] => {
    return formData.getAll("methodSteps").map(methodStep => {
      return {
        text: methodStep as string,
      };
    });
  };

  const mapToPortionSize = (formData: FormData) : PortionSize => {
    const portionSizeInput = formData?.get("portionSize")?.toString();
    if (!portionSizeInput) {
      return {
        value: "",
        measure: "",
      };
    }

    const numberIndex = portionSizeInput.search(/\d/g);
    if (numberIndex === -1) {
      return {
        value: "",
        measure: "",
      };
    }

    const nonNumberIndex = portionSizeInput.search(/\D/g);
    if (nonNumberIndex === -1 || nonNumberIndex < numberIndex) {
      return {
        value: portionSizeInput.slice(numberIndex),
        measure: "",
      };
    }

    return {
      value: portionSizeInput.slice(numberIndex, nonNumberIndex).trim(),
      measure: portionSizeInput.slice(nonNumberIndex).trim()
    };
  }

  let recipe : Recipe = {
    name: formData.get("name") as string,
    description: formData.get("description") as string,
    ingredients: mapToIngredients(formData),
    methodSteps: mapToSteps(formData),
    cookingTime: Number(formData.get("cookingTime")),
    portionSize: mapToPortionSize(formData)
  };

  try {
    const createdRecipe = await fetchFromServer<Recipe>({
      endpoint: "/kitchen/recipe",
      method: "POST",
      body: recipe,
    });
    if (createdRecipe) {
      redirect(`/my-recipes/${recipe.id}/`);
    }
  } catch (error) {
    if (error instanceof ApiError) {
      const response = error.response;

      // TODO cover other possible client errors
      if (response.status === 422) {
        const body = await response.json();
        const fields : FieldError[] = body.fields;
        return fields.map((field) => {
          const name = field.name;
          const message = field.message;
          const indexOfDot = name.indexOf(".");
          if (indexOfDot === -1) {
            return {
              name,
              message
            }
          } else {
            const mainName = name.slice(0, indexOfDot);
            return {
              name: mainName,
              message
            };
          }
        });
      }

      return [];
    }

    throw error;
  }

  throw new Error("Couldn't process the request")
}