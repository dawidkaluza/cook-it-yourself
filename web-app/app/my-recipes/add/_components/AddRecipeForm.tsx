"use client";

import {addRecipe} from "@/app/my-recipes/add/actions";
import {NameInput} from "@/app/my-recipes/add/_components/NameInput";
import {DescriptionInput} from "@/app/my-recipes/add/_components/DescriptionInput";
import {IngredientsInput} from "@/app/my-recipes/add/_components/IngredientsInput";
import {MethodStepsInput} from "@/app/my-recipes/add/_components/MethodStepsInput";
import {CookingTimeInput} from "@/app/my-recipes/add/_components/CookingTimeInput";
import {PortionSizeInput} from "@/app/my-recipes/add/_components/PortionSizeInput";
import {FormButtons} from "@/app/my-recipes/add/_components/FormButtons";
import {useFormState} from "react-dom";

const AddRecipeForm = () => {
  const [fieldErrors, action] = useFormState(addRecipe, []);

  return (
    <form action={action} noValidate className="mt-4">
      <NameInput fieldErrors={fieldErrors} />
      <DescriptionInput fieldErrors={fieldErrors} />
      <IngredientsInput />
      <MethodStepsInput />
      <CookingTimeInput />
      <PortionSizeInput />
      <FormButtons/>
    </form>
  );
};

export {AddRecipeForm};