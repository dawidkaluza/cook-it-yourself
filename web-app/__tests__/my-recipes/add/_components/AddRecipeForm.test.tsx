import {describe, expect, Mock, test, vi} from "vitest";
import {useFormState} from "react-dom";
import {render, screen} from "@testing-library/react";
import {AddRecipeForm} from "@/app/my-recipes/add/_components/AddRecipeForm";
import {FieldError} from "@/app/my-recipes/add/actions";

vi.mock("react-dom", () => {
  return {
    useFormState: vi.fn(),
  };
});

describe("AddRecipeForm component", () => {
  test("render empty form", () => {
    // Given
    (useFormState as Mock).mockReturnValue([
      [],
      () => {}
    ]);

    // When
    const formComponent = render(<AddRecipeForm />);

    // Then
    const nameInput = screen.queryByLabelText("Name");
    expect(nameInput).not.toBeNull();

    const descriptionInput = screen.queryByLabelText("Description");
    expect(descriptionInput).not.toBeNull();

    const ingredientsInput = screen.queryByLabelText("Ingredients");
    expect(ingredientsInput).not.toBeNull();

    const stepsInput = screen.queryByLabelText("Method steps");
    expect(stepsInput).not.toBeNull();

    const cookingTimeInput = screen.queryByLabelText("Cooking time");
    expect(cookingTimeInput).not.toBeNull();

    const portionSizeInput = screen.queryByLabelText("Portion size");
    expect(portionSizeInput).not.toBeNull();

    const submitButton = screen.queryByRole("button", { name: "Submit" });
    expect(submitButton).not.toBeNull();

    expect(formComponent.container).toMatchSnapshot();
    formComponent.unmount();
  });

  test.each([
    // Case 1
    [
      [
        {
          name: "name",
          message: "Some name error."
        },
      ]
    ],

    // Case 2
    [
      [
        {
          name: "description",
          message: "Some description error."
        },
        {
          name: "ingredient[0].name",
          message: "Some ingredient error."
        },
        {
          name: "methodStep[1].text",
          message: "Some method step error."
        },
      ],
    ],

    // Case 3
    [
      [
        {
          name: "cookingTime",
          message: "Some cooking time error."
        },
        {
          name: "portionSize",
          message: "Some portion size error."
        },
      ]
    ]
  ])("render with various errors", (fieldErrors: FieldError[]) => {
    // Given
    (useFormState as Mock).mockReturnValue([
      fieldErrors,
      () => {}
    ]);

    // When
    const formComponent = render(<AddRecipeForm />);

    // Then
    for (const fieldError of fieldErrors) {
      const errorMessage = screen.queryByText(fieldError.message);
      expect(errorMessage).not.toBeNull();
    }
    expect(formComponent.container).toMatchSnapshot();
    formComponent.unmount();
  });
});