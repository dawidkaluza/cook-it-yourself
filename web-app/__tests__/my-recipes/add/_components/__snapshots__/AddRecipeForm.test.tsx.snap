// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AddRecipeForm component > render empty form 1`] = `
<div>
  <form
    class="mt-4"
    novalidate=""
  >
    <div
      class="row mb-4"
    >
      <label
        class="col-sm-2 col-form-label"
        for="name"
      >
        Name
      </label>
      <div
        class="col-sm-10"
      >
        <input
          class="form-control"
          id="name"
          name="name"
          placeholder="Name of your new recipe"
        />
      </div>
    </div>
    <div
      class="row mb-4"
    >
      <label
        class="col-sm-2 col-form-label"
        for="description"
      >
        Description
      </label>
      <div
        class="col-sm-10"
      >
        <textarea
          class="form-control"
          id="description"
          name="description"
          placeholder="A few words describing your recipe"
          style="height: 100px;"
        />
      </div>
    </div>
    <div
      class="row mb-4"
    >
      <label
        class="col-sm-2 col-form-label"
        for="newIngredient"
      >
        Ingredients
      </label>
      <div
        class="col-sm-10"
      >
        <div
          class="row"
        >
          <div
            class="input-group"
          >
            <input
              class="form-control"
              id="newIngredient"
              name="newIngredient"
              placeholder="Name and optionally amount (paprikas 3, sugar 250g, etc.)"
              value=""
            />
            <button
              class="btn btn-success"
              type="button"
            >
              <svg
                class="bi bi-plus-lg"
                fill="currentColor"
                height="16"
                viewBox="0 0 16 16"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M8 2a.5.5 0 0 1 .5.5v5h5a.5.5 0 0 1 0 1h-5v5a.5.5 0 0 1-1 0v-5h-5a.5.5 0 0 1 0-1h5v-5A.5.5 0 0 1 8 2"
                  fill-rule="evenodd"
                />
              </svg>
            </button>
          </div>
        </div>
        <div
          class="row"
        />
      </div>
    </div>
    <div
      class="row mb-4"
    >
      <label
        class="col-sm-2 col-form-label"
        for="newMethodStep"
      >
        Method steps
      </label>
      <div
        class="col-sm-10"
      >
        <div
          class="row"
        >
          <div
            class="input-group"
          >
            <textarea
              class="form-control"
              id="newMethodStep"
              name="newMethodStep"
              placeholder="Instructions step-by-step to cook your recipe"
              style="height: 100px;"
            />
            <div
              class="input-group-text"
            >
              <button
                class="btn btn-sm btn-success"
                type="button"
              >
                <svg
                  class="bi bi-plus-lg"
                  fill="currentColor"
                  height="16"
                  viewBox="0 0 16 16"
                  width="16"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M8 2a.5.5 0 0 1 .5.5v5h5a.5.5 0 0 1 0 1h-5v5a.5.5 0 0 1-1 0v-5h-5a.5.5 0 0 1 0-1h5v-5A.5.5 0 0 1 8 2"
                    fill-rule="evenodd"
                  />
                </svg>
              </button>
            </div>
          </div>
        </div>
        <div
          class="row"
        />
      </div>
    </div>
    <div
      class="row mb-4"
    >
      <label
        class="col-sm-2 col-form-label"
        for="cookingTime"
      >
        Cooking time
      </label>
      <div
        class="col-sm-10"
      >
        <input
          class="form-control"
          id="cookingTime"
          name="cookingTime"
          placeholder="In minutes"
          type="number"
        />
      </div>
    </div>
    <div
      class="row mb-4"
    >
      <label
        class="col-sm-2 col-form-label"
        for="portionSize"
      >
        Portion size
      </label>
      <div
        class="col-sm-10"
      >
        <input
          class="form-control"
          id="portionSize"
          name="portionSize"
          placeholder="Amount and unit (4 plates, 800g, etc.)"
        />
      </div>
    </div>
    <div
      class="row mb-4"
    >
      <div
        class="col-12 text-center"
      >
        <div
          class="btn-group"
        >
          <button
            class="btn btn-primary"
            type="submit"
          >
            <svg
              class="bi bi-check2"
              fill="currentColor"
              height="16"
              viewBox="0 0 16 16"
              width="16"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M13.854 3.646a.5.5 0 0 1 0 .708l-7 7a.5.5 0 0 1-.708 0l-3.5-3.5a.5.5 0 1 1 .708-.708L6.5 10.293l6.646-6.647a.5.5 0 0 1 .708 0"
              />
            </svg>
             Submit 
          </button>
          <button
            class="btn btn-outline-secondary"
            type="reset"
          >
            <svg
              class="bi bi-x"
              fill="currentColor"
              height="16"
              viewBox="0 0 16 16"
              width="16"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M4.646 4.646a.5.5 0 0 1 .708 0L8 7.293l2.646-2.647a.5.5 0 0 1 .708.708L8.707 8l2.647 2.646a.5.5 0 0 1-.708.708L8 8.707l-2.646 2.647a.5.5 0 0 1-.708-.708L7.293 8 4.646 5.354a.5.5 0 0 1 0-.708"
              />
            </svg>
             Reset 
          </button>
        </div>
      </div>
    </div>
  </form>
</div>
`;

exports[`AddRecipeForm component > render with various errors 1`] = `
<div>
  <form
    class="mt-4"
    novalidate=""
  >
    <div
      class="row mb-4"
    >
      <label
        class="col-sm-2 col-form-label"
        for="name"
      >
        Name
      </label>
      <div
        class="col-sm-10"
      >
        <input
          class="form-control"
          id="name"
          name="name"
          placeholder="Name of your new recipe"
        />
        <div
          class="invalid-feedback d-block"
        >
          Some name error.
        </div>
      </div>
    </div>
    <div
      class="row mb-4"
    >
      <label
        class="col-sm-2 col-form-label"
        for="description"
      >
        Description
      </label>
      <div
        class="col-sm-10"
      >
        <textarea
          class="form-control"
          id="description"
          name="description"
          placeholder="A few words describing your recipe"
          style="height: 100px;"
        />
      </div>
    </div>
    <div
      class="row mb-4"
    >
      <label
        class="col-sm-2 col-form-label"
        for="newIngredient"
      >
        Ingredients
      </label>
      <div
        class="col-sm-10"
      >
        <div
          class="row"
        >
          <div
            class="input-group"
          >
            <input
              class="form-control"
              id="newIngredient"
              name="newIngredient"
              placeholder="Name and optionally amount (paprikas 3, sugar 250g, etc.)"
              value=""
            />
            <button
              class="btn btn-success"
              type="button"
            >
              <svg
                class="bi bi-plus-lg"
                fill="currentColor"
                height="16"
                viewBox="0 0 16 16"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M8 2a.5.5 0 0 1 .5.5v5h5a.5.5 0 0 1 0 1h-5v5a.5.5 0 0 1-1 0v-5h-5a.5.5 0 0 1 0-1h5v-5A.5.5 0 0 1 8 2"
                  fill-rule="evenodd"
                />
              </svg>
            </button>
          </div>
        </div>
        <div
          class="row"
        />
      </div>
    </div>
    <div
      class="row mb-4"
    >
      <label
        class="col-sm-2 col-form-label"
        for="newMethodStep"
      >
        Method steps
      </label>
      <div
        class="col-sm-10"
      >
        <div
          class="row"
        >
          <div
            class="input-group"
          >
            <textarea
              class="form-control"
              id="newMethodStep"
              name="newMethodStep"
              placeholder="Instructions step-by-step to cook your recipe"
              style="height: 100px;"
            />
            <div
              class="input-group-text"
            >
              <button
                class="btn btn-sm btn-success"
                type="button"
              >
                <svg
                  class="bi bi-plus-lg"
                  fill="currentColor"
                  height="16"
                  viewBox="0 0 16 16"
                  width="16"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M8 2a.5.5 0 0 1 .5.5v5h5a.5.5 0 0 1 0 1h-5v5a.5.5 0 0 1-1 0v-5h-5a.5.5 0 0 1 0-1h5v-5A.5.5 0 0 1 8 2"
                    fill-rule="evenodd"
                  />
                </svg>
              </button>
            </div>
          </div>
        </div>
        <div
          class="row"
        />
      </div>
    </div>
    <div
      class="row mb-4"
    >
      <label
        class="col-sm-2 col-form-label"
        for="cookingTime"
      >
        Cooking time
      </label>
      <div
        class="col-sm-10"
      >
        <input
          class="form-control"
          id="cookingTime"
          name="cookingTime"
          placeholder="In minutes"
          type="number"
        />
      </div>
    </div>
    <div
      class="row mb-4"
    >
      <label
        class="col-sm-2 col-form-label"
        for="portionSize"
      >
        Portion size
      </label>
      <div
        class="col-sm-10"
      >
        <input
          class="form-control"
          id="portionSize"
          name="portionSize"
          placeholder="Amount and unit (4 plates, 800g, etc.)"
        />
      </div>
    </div>
    <div
      class="row mb-4"
    >
      <div
        class="col-12 text-center"
      >
        <div
          class="btn-group"
        >
          <button
            class="btn btn-primary"
            type="submit"
          >
            <svg
              class="bi bi-check2"
              fill="currentColor"
              height="16"
              viewBox="0 0 16 16"
              width="16"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M13.854 3.646a.5.5 0 0 1 0 .708l-7 7a.5.5 0 0 1-.708 0l-3.5-3.5a.5.5 0 1 1 .708-.708L6.5 10.293l6.646-6.647a.5.5 0 0 1 .708 0"
              />
            </svg>
             Submit 
          </button>
          <button
            class="btn btn-outline-secondary"
            type="reset"
          >
            <svg
              class="bi bi-x"
              fill="currentColor"
              height="16"
              viewBox="0 0 16 16"
              width="16"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M4.646 4.646a.5.5 0 0 1 .708 0L8 7.293l2.646-2.647a.5.5 0 0 1 .708.708L8.707 8l2.647 2.646a.5.5 0 0 1-.708.708L8 8.707l-2.646 2.647a.5.5 0 0 1-.708-.708L7.293 8 4.646 5.354a.5.5 0 0 1 0-.708"
              />
            </svg>
             Reset 
          </button>
        </div>
      </div>
    </div>
  </form>
</div>
`;

exports[`AddRecipeForm component > render with various errors 2`] = `
<div>
  <form
    class="mt-4"
    novalidate=""
  >
    <div
      class="row mb-4"
    >
      <label
        class="col-sm-2 col-form-label"
        for="name"
      >
        Name
      </label>
      <div
        class="col-sm-10"
      >
        <input
          class="form-control"
          id="name"
          name="name"
          placeholder="Name of your new recipe"
        />
      </div>
    </div>
    <div
      class="row mb-4"
    >
      <label
        class="col-sm-2 col-form-label"
        for="description"
      >
        Description
      </label>
      <div
        class="col-sm-10"
      >
        <textarea
          class="form-control"
          id="description"
          name="description"
          placeholder="A few words describing your recipe"
          style="height: 100px;"
        />
        <div
          class="invalid-feedback d-block"
        >
          Some description error.
        </div>
      </div>
    </div>
    <div
      class="row mb-4"
    >
      <label
        class="col-sm-2 col-form-label"
        for="newIngredient"
      >
        Ingredients
      </label>
      <div
        class="col-sm-10"
      >
        <div
          class="row"
        >
          <div
            class="input-group"
          >
            <input
              class="form-control"
              id="newIngredient"
              name="newIngredient"
              placeholder="Name and optionally amount (paprikas 3, sugar 250g, etc.)"
              value=""
            />
            <button
              class="btn btn-success"
              type="button"
            >
              <svg
                class="bi bi-plus-lg"
                fill="currentColor"
                height="16"
                viewBox="0 0 16 16"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M8 2a.5.5 0 0 1 .5.5v5h5a.5.5 0 0 1 0 1h-5v5a.5.5 0 0 1-1 0v-5h-5a.5.5 0 0 1 0-1h5v-5A.5.5 0 0 1 8 2"
                  fill-rule="evenodd"
                />
              </svg>
            </button>
          </div>
        </div>
        <div
          class="row"
        >
          <div
            class="invalid-feedback d-block"
          >
            Some ingredient error.
          </div>
        </div>
      </div>
    </div>
    <div
      class="row mb-4"
    >
      <label
        class="col-sm-2 col-form-label"
        for="newMethodStep"
      >
        Method steps
      </label>
      <div
        class="col-sm-10"
      >
        <div
          class="row"
        >
          <div
            class="input-group"
          >
            <textarea
              class="form-control"
              id="newMethodStep"
              name="newMethodStep"
              placeholder="Instructions step-by-step to cook your recipe"
              style="height: 100px;"
            />
            <div
              class="input-group-text"
            >
              <button
                class="btn btn-sm btn-success"
                type="button"
              >
                <svg
                  class="bi bi-plus-lg"
                  fill="currentColor"
                  height="16"
                  viewBox="0 0 16 16"
                  width="16"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M8 2a.5.5 0 0 1 .5.5v5h5a.5.5 0 0 1 0 1h-5v5a.5.5 0 0 1-1 0v-5h-5a.5.5 0 0 1 0-1h5v-5A.5.5 0 0 1 8 2"
                    fill-rule="evenodd"
                  />
                </svg>
              </button>
            </div>
          </div>
        </div>
        <div
          class="row"
        >
          <div
            class="invalid-feedback d-block"
          >
            Some method step error.
          </div>
        </div>
      </div>
    </div>
    <div
      class="row mb-4"
    >
      <label
        class="col-sm-2 col-form-label"
        for="cookingTime"
      >
        Cooking time
      </label>
      <div
        class="col-sm-10"
      >
        <input
          class="form-control"
          id="cookingTime"
          name="cookingTime"
          placeholder="In minutes"
          type="number"
        />
      </div>
    </div>
    <div
      class="row mb-4"
    >
      <label
        class="col-sm-2 col-form-label"
        for="portionSize"
      >
        Portion size
      </label>
      <div
        class="col-sm-10"
      >
        <input
          class="form-control"
          id="portionSize"
          name="portionSize"
          placeholder="Amount and unit (4 plates, 800g, etc.)"
        />
      </div>
    </div>
    <div
      class="row mb-4"
    >
      <div
        class="col-12 text-center"
      >
        <div
          class="btn-group"
        >
          <button
            class="btn btn-primary"
            type="submit"
          >
            <svg
              class="bi bi-check2"
              fill="currentColor"
              height="16"
              viewBox="0 0 16 16"
              width="16"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M13.854 3.646a.5.5 0 0 1 0 .708l-7 7a.5.5 0 0 1-.708 0l-3.5-3.5a.5.5 0 1 1 .708-.708L6.5 10.293l6.646-6.647a.5.5 0 0 1 .708 0"
              />
            </svg>
             Submit 
          </button>
          <button
            class="btn btn-outline-secondary"
            type="reset"
          >
            <svg
              class="bi bi-x"
              fill="currentColor"
              height="16"
              viewBox="0 0 16 16"
              width="16"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M4.646 4.646a.5.5 0 0 1 .708 0L8 7.293l2.646-2.647a.5.5 0 0 1 .708.708L8.707 8l2.647 2.646a.5.5 0 0 1-.708.708L8 8.707l-2.646 2.647a.5.5 0 0 1-.708-.708L7.293 8 4.646 5.354a.5.5 0 0 1 0-.708"
              />
            </svg>
             Reset 
          </button>
        </div>
      </div>
    </div>
  </form>
</div>
`;

exports[`AddRecipeForm component > render with various errors 3`] = `
<div>
  <form
    class="mt-4"
    novalidate=""
  >
    <div
      class="row mb-4"
    >
      <label
        class="col-sm-2 col-form-label"
        for="name"
      >
        Name
      </label>
      <div
        class="col-sm-10"
      >
        <input
          class="form-control"
          id="name"
          name="name"
          placeholder="Name of your new recipe"
        />
      </div>
    </div>
    <div
      class="row mb-4"
    >
      <label
        class="col-sm-2 col-form-label"
        for="description"
      >
        Description
      </label>
      <div
        class="col-sm-10"
      >
        <textarea
          class="form-control"
          id="description"
          name="description"
          placeholder="A few words describing your recipe"
          style="height: 100px;"
        />
      </div>
    </div>
    <div
      class="row mb-4"
    >
      <label
        class="col-sm-2 col-form-label"
        for="newIngredient"
      >
        Ingredients
      </label>
      <div
        class="col-sm-10"
      >
        <div
          class="row"
        >
          <div
            class="input-group"
          >
            <input
              class="form-control"
              id="newIngredient"
              name="newIngredient"
              placeholder="Name and optionally amount (paprikas 3, sugar 250g, etc.)"
              value=""
            />
            <button
              class="btn btn-success"
              type="button"
            >
              <svg
                class="bi bi-plus-lg"
                fill="currentColor"
                height="16"
                viewBox="0 0 16 16"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M8 2a.5.5 0 0 1 .5.5v5h5a.5.5 0 0 1 0 1h-5v5a.5.5 0 0 1-1 0v-5h-5a.5.5 0 0 1 0-1h5v-5A.5.5 0 0 1 8 2"
                  fill-rule="evenodd"
                />
              </svg>
            </button>
          </div>
        </div>
        <div
          class="row"
        />
      </div>
    </div>
    <div
      class="row mb-4"
    >
      <label
        class="col-sm-2 col-form-label"
        for="newMethodStep"
      >
        Method steps
      </label>
      <div
        class="col-sm-10"
      >
        <div
          class="row"
        >
          <div
            class="input-group"
          >
            <textarea
              class="form-control"
              id="newMethodStep"
              name="newMethodStep"
              placeholder="Instructions step-by-step to cook your recipe"
              style="height: 100px;"
            />
            <div
              class="input-group-text"
            >
              <button
                class="btn btn-sm btn-success"
                type="button"
              >
                <svg
                  class="bi bi-plus-lg"
                  fill="currentColor"
                  height="16"
                  viewBox="0 0 16 16"
                  width="16"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M8 2a.5.5 0 0 1 .5.5v5h5a.5.5 0 0 1 0 1h-5v5a.5.5 0 0 1-1 0v-5h-5a.5.5 0 0 1 0-1h5v-5A.5.5 0 0 1 8 2"
                    fill-rule="evenodd"
                  />
                </svg>
              </button>
            </div>
          </div>
        </div>
        <div
          class="row"
        />
      </div>
    </div>
    <div
      class="row mb-4"
    >
      <label
        class="col-sm-2 col-form-label"
        for="cookingTime"
      >
        Cooking time
      </label>
      <div
        class="col-sm-10"
      >
        <input
          class="form-control"
          id="cookingTime"
          name="cookingTime"
          placeholder="In minutes"
          type="number"
        />
        <div
          class="invalid-feedback d-block"
        >
          Some cooking time error.
        </div>
      </div>
    </div>
    <div
      class="row mb-4"
    >
      <label
        class="col-sm-2 col-form-label"
        for="portionSize"
      >
        Portion size
      </label>
      <div
        class="col-sm-10"
      >
        <input
          class="form-control"
          id="portionSize"
          name="portionSize"
          placeholder="Amount and unit (4 plates, 800g, etc.)"
        />
        <div
          class="invalid-feedback d-block"
        >
          Some portion size error.
        </div>
      </div>
    </div>
    <div
      class="row mb-4"
    >
      <div
        class="col-12 text-center"
      >
        <div
          class="btn-group"
        >
          <button
            class="btn btn-primary"
            type="submit"
          >
            <svg
              class="bi bi-check2"
              fill="currentColor"
              height="16"
              viewBox="0 0 16 16"
              width="16"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M13.854 3.646a.5.5 0 0 1 0 .708l-7 7a.5.5 0 0 1-.708 0l-3.5-3.5a.5.5 0 1 1 .708-.708L6.5 10.293l6.646-6.647a.5.5 0 0 1 .708 0"
              />
            </svg>
             Submit 
          </button>
          <button
            class="btn btn-outline-secondary"
            type="reset"
          >
            <svg
              class="bi bi-x"
              fill="currentColor"
              height="16"
              viewBox="0 0 16 16"
              width="16"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M4.646 4.646a.5.5 0 0 1 .708 0L8 7.293l2.646-2.647a.5.5 0 0 1 .708.708L8.707 8l2.647 2.646a.5.5 0 0 1-.708.708L8 8.707l-2.646 2.647a.5.5 0 0 1-.708-.708L7.293 8 4.646 5.354a.5.5 0 0 1 0-.708"
              />
            </svg>
             Reset 
          </button>
        </div>
      </div>
    </div>
  </form>
</div>
`;
